### PR TITLE
pageserver: only throttle pagestream requests & bring back throttling deduction for smgr latency metrics

### PR DESCRIPTION
**Pull request by problame**
## Problem

In the batching PR 
- https://github.com/neondatabase/neon/pull/9870

I stopped deducting the time-spent-in-throttle fro latency metrics, i.e.,
- smgr latency metrics (`SmgrOpTimer`)
- basebackup latency (+scan latency, which I think is part of basebackup).

The reason for stopping the deduction was that with the introduction of batching, the trick with tracking time-spent-in-throttle inside RequestContext and swap-replacing it from the `impl Drop for SmgrOpTimer` no longer worked with >1 requests in a batch.

However, deducting time-spent-in-throttle is desirable because our internal latency SLO definition does not account for throttling.

## Summary of changes

- Redefine throttling to be a page_service pagestream request throttle instead of a throttle for repository `Key` reads through `Timeline::get` / `Timeline::get_vectored`.
  - This means reads done by `basebackup` are no longer subject to any throttle.
- The throttle applies after batching, before handling of the request.
- Drive-by fix: make throttle sensitive to cancellation.
- Rename metric label `kind` from `timeline_get` to `pagestream` to reflect the new scope of throttling.

To avoid config format breakage, we leave the config field named `timeline_get_throttle` and ignore the `task_kinds` field.
This will be cleaned up in a future PR.

## Trade-Offs

Ideally, we would apply the throttle before reading a request off the connection, so that we queue the minimal amount of work inside the process.
However, that's not possible because we need to do shard routing.

The redefinition of the throttle to limit pagestream request rate instead of repository `Key` rate comes with several downsides:
- We're no longer able to use the throttle mechanism for other other tasks, e.g. image layer creation.
  However, in practice, we never used that capability anyways.
- We no longer throttle basebackup.

### CODE REVIEW

File: pageserver/src/page_service.rs [hunk 1]
```diff
@@ -574,6 +574,41 @@ enum BatchedFeMessage {
...
+            | BatchedFeMessage::DbSize { shard, timer, .. }
+            | BatchedFeMessage::GetSlruSegment { shard, timer, .. } => {
+                (
+                    shard,
+                    // 1 token is probably under-estimating because these
+                    // request handlers typically do several Timeline::get calls.
+                    1,
+                    itertools::Either::Left(std::iter::once(timer)),
+                )
+            }
```
Comment (VladLazar) [inline pageserver/src/page_service.rs:591]: Discussed this out of band, but leaving it here too for completeness:

`BatchedFeMessage::Exists` and `BatchedFeMessage::Nblocks` may be served from the relsize cache and not materialize into a get. Should we try to account for this?

**Review by VladLazar — COMMENTED** [review]
Code looks fine, but I don't get the test failures. Possibly related to my comment, but not sure.

File: pageserver/src/page_service.rs (hunk 1 above)
Comment (problame) [reply pageserver/src/page_service.rs:591]: https://github.com/neondatabase/neon/pull/9962/commits/d679883a085ff411e6c35cfbb5a3a6ca861ef731

**Review by VladLazar — APPROVED** [review]

File: pageserver/src/page_service.rs [hunk 2]
```diff
@@ -574,6 +574,41 @@ enum BatchedFeMessage {
...
+                itertools::Either::Right(pages.iter_mut().map(|(_, _, timer)| timer)),
+            ),
+            BatchedFeMessage::RespondError { .. } => return Ok(()),
+        };
+        let throttled = tokio::select! {
+            throttled = shard.pagestream_throttle.throttle(tokens) => { throttled }
+            _ = cancel.cancelled() => {
+                return Err(QueryError::Shutdown);
+            }
+        };
```
Comment (problame) [inline pageserver/src/page_service.rs:604]: `throttle()` is not actually cancel-safe.

For example, accounted_started vs accounted_finished metrics will get out of balance

**Pull request merged**